redskyctl/internal/commands/generate/rbac.go

```diff
@@ -90,88 +96,174 @@ func (o *RBACOptions) Complete() {
-	if len(clusterRole.Rules) == 0 {
-		return nil
-	}
+	// Namespaces
+	var namespaces []string
+	if !o.ClusterRole || !o.ClusterRoleBinding {
+		// Get the distinct list of namespaces from the experiments
+		ns := make(map[string]bool, len(experimentList.Items))
+		for i := range experimentList.Items {
+			ns[experimentList.Items[i].Namespace] = true
+		}
 
-	l := &corev1.List{
-		Items: []runtime.RawExtension{
-			{Object: clusterRole},
-			{Object: clusterRoleBinding},
-		},
+		// Try to provide an explicit default value (cstr.Namespace can still be "")
+		if ns[""] {
+			if cstr, err := config.CurrentCluster(o.Config.Reader()); err == nil {
+				delete(ns, "")
+				ns[cstr.Namespace] = true
+			}
+		}
+
+		namespaces = make([]string, 0, len(ns))
+		for k := range ns {
+			namespaces = append(namespaces, k)
+		}
 	}
-	return o.Printer.PrintObj(l, o.Out)
+
+	// The manager runs as the default service account
+	return roleRef, subject, namespaces, nil
 }
 
-// findRules finds the patch targets from an experiment
-func (o *RBACOptions) findRules(exp *redskyv1alpha1.Experiment) ([]*rbacv1.PolicyRule, error) {
-	var rules []*rbacv1.PolicyRule
+func buildRBAC(roleRef *rbacv1.RoleRef, subject *rbacv1.Subject, rules []rbacv1.PolicyRule, namespaces []string) *corev1.List {
+	result := &corev1.List{}
+	switch roleRef.Kind {
+
+	case "ClusterRole":
+		// Include a single cluster role
+		result.Items = append(result.Items, runtime.RawExtension{
+			Object: &rbacv1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: roleRef.Name,
+				},
+				Rules: rules,
+			},
+		})
+
+		// For each namespace, include a role binding
+		for _, ns := range namespaces {
+			result.Items = append(result.Items, runtime.RawExtension{
+				Object: &rbacv1.RoleBinding{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      roleRef.Name + "binding",
+						Namespace: ns,
+					},
+					Subjects: []rbacv1.Subject{*subject},
+					RoleRef:  *roleRef,
+				},
+			})
+		}
+
+		// If there are no namespaces, include a single cluster role binding
+		if len(namespaces) == 0 {
+			result.Items = append(result.Items, runtime.RawExtension{
+				Object: &rbacv1.ClusterRoleBinding{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: roleRef.Name + "binding",
+					},
+					Subjects: []rbacv1.Subject{*subject},
+					RoleRef:  *roleRef,
+				},
+			})
+		}
+
+	case "Role":
+		for _, ns := range namespaces {
+			// Include a role and role binding for each namespace
+			result.Items = append(result.Items,
+				runtime.RawExtension{
+					Object: &rbacv1.Role{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      roleRef.Name,
+							Namespace: ns,
+						},
+						Rules: rules,
+					},
+				},
+				runtime.RawExtension{
+					Object: &rbacv1.RoleBinding{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      roleRef.Name + "binding",
+							Namespace: ns,
+						},
+						Subjects: []rbacv1.Subject{*subject},
+						RoleRef:  *roleRef,
+					},
+				})
+		}
+	}
+	return result
+}
```

Comment:
What about moving the role binding generation out of the switch statement since it's common in both role and cluster role? -- 


```
	case "ClusterRole":
		// Include a single cluster role
		result.Items = append(result.Items, runtime.RawExtension{
			Object: &rbacv1.ClusterRole{
				ObjectMeta: metav1.ObjectMeta{
					Name: roleRef.Name,
				},
				Rules: rules,
			},
		})

		if len(namespaces) > 0 {
			break
		}
		
		// If there are no namespaces, include a single cluster role binding
		result.Items = append(result.Items, runtime.RawExtension{
			Object: &rbacv1.ClusterRoleBinding{
				ObjectMeta: metav1.ObjectMeta{
					Name: roleRef.Name + "binding",
				},
				Subjects: []rbacv1.Subject{*subject},
				RoleRef:  *roleRef,
			},
		})

	case "Role":
		for _, ns := range namespaces {
			// Include a role and role binding for each namespace
			result.Items = append(result.Items,
				runtime.RawExtension{
					Object: &rbacv1.Role{
						ObjectMeta: metav1.ObjectMeta{
							Name:      roleRef.Name,
							Namespace: ns,
						},
						Rules: rules,
					},
				})
		}
	}

	// For each namespace, include a role binding
	for _, ns := range namespaces {
		result.Items = append(result.Items, runtime.RawExtension{
			Object: &rbacv1.RoleBinding{
				ObjectMeta: metav1.ObjectMeta{
					Name:      roleRef.Name + "binding",
					Namespace: ns,
				},
				Subjects: []rbacv1.Subject{*subject},
				RoleRef:  *roleRef,
			},
		})
	}

	return result
```

Reply:
So much better.